src/Proxy.NET/Services/CatalogService.cs

```diff
@@ -0,0 +1,109 @@
+using System.Data;
+using System.Net;
+using AzureOpenAIProxy.Management;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Caching.Memory;
+using Npgsql;
+using Proxy.NET.Models;
+
+namespace Proxy.NET.Services;
+
+public class CatalogService(AoaiProxyContext db, IConfiguration configuration, IMemoryCache memoryCache) : ICatalogService
+{
+    private readonly string EncryptionKey =
+        configuration["PostgresEncryptionKey"] ?? throw new ArgumentNullException("PostgresEncryptionKey");
+
+    /// <summary>
+    /// Retrieves the event catalog for a given event ID and deployment name.
+    /// Calls function aoai.get_models_by_deployment_name that decrypts the endpoint URL and Key.
+    /// </summary>
+    /// <param name="eventId">The ID of the event.</param>
+    /// <param name="deploymentName">The name of the deployment.</param>
+    /// <returns>A list of Deployment objects representing the event catalog.</returns>
+    private async Task<List<Deployment>> GetDecryptedEventCatalogAsync(string eventId, string deploymentName)
+    {
+        if (memoryCache.TryGetValue(eventId + deploymentName, out List<Deployment>? cachedContext))
+            return cachedContext!;
+
+        var result = await db.Set<Deployment>()
+            .FromSqlRaw(
+                "SELECT * FROM aoai.get_models_by_deployment_name(@eventId, @deploymentName, @encryptionKey)",
+                new NpgsqlParameter("@eventId", eventId),
+                new NpgsqlParameter("@deploymentName", deploymentName),
+                new NpgsqlParameter("@encryptionKey", EncryptionKey)
+            )
+            .ToListAsync();
```

Comment:
This is where stored procs start to fall down. We should have this in a much more type-safe approach with LINQ rather than having to use a stored proc

Reply:
But it is type safe <Deployment>. Also, this function does two decryptions - one for the key and the other for the deployment. That will result in 3 calls to PG, where as this function achieves with just one call

Reply:
It's not _really_ type safe, as it expects that you've created the type + mapping correctly, which wasn't the case with `RequestContext`.

It's true that we should use sprocs in some spots but that's why I added [#310](https://github.com/microsoft/azure-openai-service-proxy/issues/310) as we might be overdoing it now.

Reply:
I have dropped the SPs that weren't doing too much. But the others all perform more complex multi step inserts and updates